// textlint-plugin-latex2e/lib/LaTeXProcessor.js
"use strict";
/*
 * Copyright (c) 2020 Textlint Plugin LaTex2e team. All rights reserved.
 *
 * This file is part of textlint-plugin-latex2e.
 * This software is released under the MIT License, see LICENSE.md .
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaTeXProcessor = void 0;
const latex_to_ast_1 = require("./latex-to-ast");
class LaTeXProcessor {
    constructor(config = {}) {
        this.config = config;
    }
    availableExtensions() {
        return [".tex", ".cls"].concat(this.config.extensions || []);
    }
    processor() {
        return {
            preProcess(text) {
                return latex_to_ast_1.parse(text);
            },
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            postProcess(messages, filePath) {
                return {
                    messages,
                    filePath: filePath ? filePath : "<latex>",
                };
            },
        };
    }
}
exports.LaTeXProcessor = LaTeXProcessor;
